### PR TITLE
WIP: Constrain longitude between -180 and 180

### PR DESCRIPTION
**Pull request by zachasme**
The longitude range [0,360] tripped me up. Are you still interested in pushing the [-180,180] range down to C code, cf. [this comment](https://github.com/uber/h3/issues/38#issuecomment-376670998)?

If so I will go over the test suite as well.

### CODE REVIEW

**Comment by nrabinowitz** [comment]
I don't object, but technically this might need to be a major version change, and as such might be worth putting off. Though in practice it wouldn't actually break the Java or JS bindings, it would just render their normalization functions unnecessary. Would like thoughts from @isaacbrodsky and @dfellis.

**Comment by nrabinowitz** [comment]
(I should have added, thanks for the PR! I didn't actually realize this change could be implemented in such a minimal fashion.)

**Comment by dfellis** [comment]
I'm really excited for this change, and also surprised that nothing needed to be done to the handling of latitudes, which I thought were similarly in the `[0, Pi)` range instead of `[-Pi/2, Pi/2)` range. (But then looking at the rest of the function it assumes that the north pole is `Pi/2` and the south pole at `-Pi/2`.)

Since it still *accepts* inputs greater than Pi for latitude and then converts them to the expected range, I don't think this is a backwards incompatible change (since the greater than Pi radians are equivalent to the less than 0 radians on the globe), especially for those using the bindings where it will run it through the `constrain` functions and convert to degrees from radians.

My only concerns are whether or not other parts of the code have hardwired expectations of only positive radians. The test suite shows that `geoToH3` is perfectly fine, which is great and was my biggest worry, but the failures in `testH3SetToLinkedGeo` and `testH3SetToVertexGraph` imply that there may be positive-only expectations hardwired in those features and the failure in `testH3Api` implies that the `constrainLng` call might be breaking `NAN` and/or `INFINITY` detection on the inputs.

**Comment by nrabinowitz** [comment]
The test failures for `testH3SetToLinkedGeo` and `testH3SetToVertexGraph` are segfaults, which suggests they might be fixed by #94 - it would be worth rebasing to master and seeing if those issues go away.

**Comment by isaacbrodsky** [comment]
@zachasme thanks for the PR, I think it definitely makes sense to move this into the core rather than the bindings.

You may want to rebase now that #94 is merged, since that improves support for negative coordinates. `testH3SetToVertexGraph` gets fixed, but it looks like `testH3SetToLinkedGeo` still has some failure have rebasing. I didn't look into that yet. `testH3Api` looks good to me, it's failing on a test for exact vertices, not the Infinity/NaN handling.

I don't think this requires a major version bump. Input should not be impacted and supported both negative and positive coordinates before. Output is different but is equivalent, and bindings were doing this transform before anyways.

**Comment by nrabinowitz** [comment]
Hi @zachasme - I think we're all on board for this change. Are you planning to do the test updates, or should one of the maintainers take this on?

**Comment by zachasme** [comment]
I'm planning to finish this when I get back from vacation next week.

**Comment by nrabinowitz** [comment]
> The failure in testH3SetToLinkedGeo is due to the loops being reversed, possibly related to #53?

Hm. I will look into this today.
- Changes in the order of the list of loops are fine (we will normalize in #53, but even after that the order of the holes is undefined), and can be addressed in tests
- Changes in the starting coordinate of a given loop sequence are fine (all loops are closed, so the start coord is undefined), and can be addressed in tests
- Changes in the order or direction (clockwise vs counter-clockwise) of a given loop sequence is **not** fine, and will need to be addressed in code

**Comment by nrabinowitz** [comment]
Re test failures in `testH3SetToLinkedGeo` - this is okay; the order of the loops within the list of loops is currently undefined/arbitrary, so reversing that order is acceptable.

Patch for this:

```
--- a/src/apps/testapps/testH3SetToLinkedGeo.c
+++ b/src/apps/testapps/testH3SetToLinkedGeo.c
@@ -119,10 +119,12 @@ TEST(hole) {
     H3_EXPORT(h3SetToLinkedGeo)(set, numHexes, &polygon);
 
     t_assert(countLinkedLoops(&polygon) == 2, "2 loops added to polygon");
-    t_assert(countLinkedCoords(polygon.first) == 6 * 3,
-             "All outer coords added to first loop");
-    t_assert(countLinkedCoords(polygon.first->next) == 6,
-             "All inner coords added to second loop");
+    // Note: This isn't strictly correct, and should be reversed when
+    // https://github.com/uber/h3/issues/53 is resolved
+    t_assert(countLinkedCoords(polygon.first) == 6,
+             "All inner coords added to first loop");
+    t_assert(countLinkedCoords(polygon.first->next) == 6 * 3,
+             "All outer coords added to second loop");
 
     H3_EXPORT(destroyLinkedPolygon)(&polygon);
```

**Comment by zachasme** [comment]
I have applied the patch from @nrabinowitz, and fixed the remaining tests as well.

File: src/h3lib/lib/geoCoord.c [hunk 1]
```diff
@@ -127,6 +127,9 @@ double constrainLng(double lng) {
     while (lng > M_PI) {
```
Comment (isaacbrodsky) [inline src/h3lib/lib/geoCoord.c]: Might be nice to replace this with something constant time in the future

**Pull request merged**